src/core/allow-bind.c

```diff
@@ -0,0 +1,164 @@
+
+        r = allow_bind_bpf__load(obj);
+        if (r)
+                return log_unit_error_errno(u, r, "Failed to load BPF object");
+
+        map_fd = bpf_map__fd(obj->maps.ports_v6);
+        if (map_fd < 0)
+                return log_unit_error_errno(u, map_fd, "Failed to get BPF map fd");
+
+        SET_FOREACH(port, ports_v6)
+                if (bpf_map_update_elem(map_fd, &port, &dummy, BPF_ANY))
+                        return log_unit_error_errno(u, -1, "Failed to update BPF map");
+
+        map_fd = bpf_map__fd(obj->maps.ports_v4);
+        if (map_fd < 0)
+                return log_unit_error_errno(u, map_fd, "Failed to get BPF map fd");
+
+        SET_FOREACH(port, ports_v4)
+                if (bpf_map_update_elem(map_fd, &port, &dummy, BPF_ANY))
+                        return log_unit_error_errno(u, -1, "Failed to update BPF map");
+
+        *ret_obj = TAKE_PTR(obj);
+        return 0;
+}
+
+int allow_bind_supported(void) {
+        _cleanup_(allow_bind_bpf_freep) struct allow_bind_bpf *obj = NULL;
+
+        int r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
+        if (r < 0)
+                return log_error_errno(r, "Can't determine whether the unified hierarchy is used: %m");
+
+        if (r == 0) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                "Not running with unified cgroup hierarchy, BPF is not supported");
+                return 0;
+        }
+
+        if (!bpf_probe_prog_type(BPF_PROG_TYPE_CGROUP_SOCK_ADDR, /*ifindex=*/0)) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                "BPF program type cgroup_sock_addr is not supported");
+                return 0;
+        }
+
+        r = prepare_allow_bind_bpf(NULL, NULL, NULL, &obj);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to load BPF object: %m");
+
+        return can_link_bpf_program(obj->progs.allow_bind_v6);
+}
+
+int allow_bind_install(Unit *u) {
+        _cleanup_(bpf_link_freep) struct bpf_link *ipv4 = NULL, *ipv6 = NULL;
+        _cleanup_(allow_bind_bpf_freep) struct allow_bind_bpf *obj = NULL;
+        _cleanup_free_ char *cgroup_path = NULL;
+        _cleanup_close_ int cgroup_fd = -1;
+        CGroupContext *cc;
+        int r;
+
+        cc = unit_get_cgroup_context(u);
+        if (!cc)
+                return 0;
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, NULL, &cgroup_path);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to get cgroup path: %m");
+
+        if (!cc->allow_bind_ports)
+                return 0;
+
+        r = prepare_allow_bind_bpf(u, cc->allow_bind_ports, cc->allow_bind_ports, &obj);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to load BPF object: %m");
+
+        cgroup_fd = open(cgroup_path, O_RDONLY | O_CLOEXEC, 0);
+        if (cgroup_fd < 0)
+                return -errno;
```

Comment:
should log here too.

we usually want functions that log on all error paths, or on none — in which case it's the callers responsibility to do so using the returned error. This function looks like it's of the logging kind, hence should log about this error too